src/process_utility.c

```diff
@@ -1206,12 +1236,142 @@ process_index_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
+		 * We acquire and immediately release a lock on a barrier table to do
+		 * this.
+		 */
+		Relation	barrier = relation_open(info->extended_options.barrier_table, AccessExclusiveLock);
+
+		relation_close(barrier, AccessExclusiveLock);
+	}
+#endif
+
+	/*
+	 * Change user since chunks are typically located in an internal schema
+	 * and chunk indexes require metadata changes. In the single-tranaction
+	 * case, we do this once for the entire table.
+	 */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+
+	/*
+	 * Hold a lock on the hypertable, it's index, and the chunk to prevent
+	 * from being altered. Since we use the same relids across transactions,
+	 * there is a potential issue if the id gets reassigned between one
+	 * sub-transaction and the next. CLUSTER has a similar issue.
+	 *
+	 * We grab a ShareLock on the chunk, because that's what CREATE INDEX
+	 * does. For the hypertable, and the hypertable's index, we are ok using
+	 * the weaker AccessShareLock, since we only need to prevent the index
+	 * iteself from being ALTERed or DROPed during this part of index
+	 * creation. TODO: is the above correct?
+	 */
+	hypertable_rel = relation_open(ht->main_table_relid, AccessShareLock);
+	chunk_rel = relation_open(chunk_relid, ShareLock);
+	hypertable_index_rel = relation_open(info->obj.objectId, AccessShareLock);
+
+	chunk = ts_chunk_get_by_relid(chunk_relid, ht->space->num_dimensions, true);
+
+	/*
+	 * use ts_chunk_index_create instead of ts_chunk_index_create_from_stmt to
+	 * handle cases where the index is altered. Validation happens when
+	 * creating the hypertable's index, which goes through the ususal
+	 * DefineIndex mechanism.
+	 */
+	ts_chunk_index_create(hypertable_rel, ht->fd.id, hypertable_index_rel, chunk->fd.id, chunk_rel);
+
+	relation_close(hypertable_index_rel, NoLock);
+	relation_close(chunk_rel, NoLock);
+	relation_close(hypertable_rel, NoLock);
+
+	ts_catalog_restore_user(&sec_ctx);
+}
+
+/*
+ * passing this via static var until we figure out a better way to do it,
+ * maybe reallocate parsetree?
+ */
+
+typedef enum HypertableIndexFlags
+{
+	HypertableIndexFlagMultiTransaction = 0,
+	HypertableIndexFlagBarrierTable,
+	HypertableIndexFlagMaxChunks,
+
+}			HypertableIndexFlags;
+
+static const WithClauseArg index_with_clauses[] = {
+	[HypertableIndexFlagMultiTransaction] = {.arg_name = "multi_transaction",.deserializer = ts_with_clause_deserialize_bool,},
+	[HypertableIndexFlagBarrierTable] = {.arg_name = "barrier_table",.deserializer = ts_with_clause_deserialize_regclass,},
+	[HypertableIndexFlagMaxChunks] = {.arg_name = "max_chunks",.deserializer = ts_with_clause_deserialize_int32,},
+};
+
+static HypertableIndexOptions hypertable_index_options = {0,.max_chunks = -1,};
+
+static HypertableIndexOptions hypertable_index_options_default = {0,.max_chunks = -1,};
+
+static bool
+on_index_arg(Size index, const char *name, Datum value, void *state)
```

Comment:
I originally had it like that, but I switched over to this version because:

1. We have a few common deserialization functions which basically all our WITH clauses will use, and standardizing them makes error reporting much nicer.
2. Due to the above, passing in multiple argument handling functions was actually _more_ verbose, since instead of having just the one `switch` I need a full function header for each argument; I actually ended up writing a single argument-handler which I passed in to all the WithClauseArgs to cut down on the bloat.
2. Scattering the argument handling across half a dozen functions was actually harder to read than this version, where there's a single place to look at when you want a sense of the valid arguments.